### ISSUE TITLE
9l doesn't support FreeBSD 12

### ISSUE THREAD

**Issue by francisconti** [9fans/plan9port#247]
The 'case $tag' in 9l for FreeBSD needs to be updated to deal with FreeBSD 12. Changing "1[0-1]" to "1[0-2]" is the obvious trivial way to do this, but should we really need to revisit this every major version? It seems like it'd be better to change that entire sub-case to the default (I'm not sure what the behavior of FreeBSD < 5.2 is here, but given that would be before 2004-01, I'm not sure we ought to worry too much).

Happy to make submit either change, but would appreciate guidance from the maintainers.

**Comment by jxy** [comment]
Actually do we need `-lpthread`?  It looks like `libthr` is automatically linked for me and I have never had a problem with `9l`.

**Comment by francisconti** [comment]
With stock p9p and a freshly-installed FreeBSD 12, INSTALL fails when 9l can't find some pthread symbol (pthread_create, IIRC). Changing "1[0-1]" to "1[0-2]" let installation complete.

I'm curious how you're getting libthr linked. How are you determining that?

**Comment by jxy** [comment]
I just realized that I did make the change to `9l` and I completely forgot about it.  I guess change the pattern to be `1[0-9]*` to catch all version 10 or later would be good.

**Comment by jxy** [comment]
Looking through the diffs, I also made some changes to `libmach/FreeBSD.c`
```
diff --git a/src/libmach/FreeBSD.c b/src/libmach/FreeBSD.c
index 7cf5c1f4..d6b00de5 100644
--- a/src/libmach/FreeBSD.c
+++ b/src/libmach/FreeBSD.c
@@ -30,8 +30,8 @@ struct PtraceRegs
        int pid;
 };
 
-static int ptracerw(Map*, Seg*, ulong, void*, uint, int);
-static int ptraceregrw(Regs*, char*, ulong*, int);
+static int ptracerw(Map*, Seg*, u64int, void*, uint, int);
+static int ptraceregrw(Regs*, char*, u64int*, int);
 
 void
 unmapproc(Map *map)
@@ -92,7 +92,7 @@ detachproc(int pid)
 }
 
 static int
-ptracerw(Map *map, Seg *seg, ulong addr, void *v, uint n, int isr)
+ptracerw(Map *map, Seg *seg, u64int addr, void *v, uint n, int isr)
 {
        int i;
        u32int u;
@@ -155,7 +155,7 @@ static char *freebsdregs[] = {
        "GS",
 };
 
-static ulong
+static u64int
 reg2freebsd(char *reg)
 {
        int i;
@@ -163,20 +163,20 @@ reg2freebsd(char *reg)
        for(i=0; i<nelem(freebsdregs); i++)
                if(strcmp(freebsdregs[i], reg) == 0)
                        return 4*i;
-       return ~(ulong)0;
+       return ~(u64int)0;
 }
 
 static int
-ptraceregrw(Regs *regs, char *name, ulong *val, int isr)
+ptraceregrw(Regs *regs, char *name, u64int *val, int isr)
 {
        int pid;
-       ulong addr;
+       u64int addr;
        struct reg mregs;
 
        addr = reg2freebsd(name);
        if(~addr == 0){
                if(isr){
-                       *val = ~(ulong)0;
+                       *val = ~(u64int)0;
                        return 0;
                }
                werrstr("register not available");
@@ -187,9 +187,9 @@ ptraceregrw(Regs *regs, char *name, ulong *val, int isr)
        if(ptrace(PT_GETREGS, pid, (char*)&mregs, 0) < 0)
                return -1;
        if(isr)
-               *val = *(u32int*)((char*)&mregs+addr);
+               *val = *(u64int*)((char*)&mregs+addr);
        else{
-               *(u32int*)((char*)&mregs+addr) = *val;
+               *(u64int*)((char*)&mregs+addr) = *val;
                if(ptrace(PT_SETREGS, pid, (char*)&mregs, 0) < 0)
                        return -1;
        }
```

**Comment by francisconti** [comment]
I got those warnings, too, but hadn't gotten there yet. :-) I'd suggest submitting those separately. Lacking guidance to the contrary, I'll submit a PR for the "medium" fix of 9l and we can address it again in several years when FreeBSD 20 comes out.

**Comment by rsc** [comment]
Fixed by 715807d
